File: apps/mobile/src/database/models/Message.ts
import { Model } from '@nozbe/watermelondb';
import { field, date, readonly } from '@nozbe/watermelondb/decorators';

export default class Message extends Model {
  static table = 'messages';
  static associations = {
    direct_chats: { type: 'belongs_to', key: 'chat_id' },
    group_chats: { type: 'belongs_to', key: 'chat_id' },
  } as const;

  @field('chat_id') chatId!: string;
  @field('chat_type') chatType!: string;
  @field('sender_id') senderId!: string;
  @field('sender_name') senderName!: string;
  @field('text') text?: string;
  @field('type') type!: string;
  @field('media_url') mediaUrl?: string;
  @field('location') locationRaw?: string; // JSON string
  @field('reply_to') replyToRaw?: string; // JSON string (ReplyTo object)
  @field('status') status!: string;
  @field('read_by') readByRaw?: string; // JSON string
  @field('delivered_to') deliveredToRaw?: string; // JSON string
  @field('deleted_for') deletedForRaw?: string; // JSON string
  @field('deleted_for_everyone_at') deletedForEveryoneAt?: number;
  @field('mentions') mentionsRaw?: string; // JSON string
  
  @readonly @date('created_at') createdAt!: Date;
  @readonly @date('updated_at') updatedAt!: Date;
  @field('edited_at') editedAt?: number;

  get locationData(): any {
    try {
      return this.locationRaw ? JSON.parse(this.locationRaw) : null;
    } catch {
      return null;
    }
  }

  get replyToData(): any {
    try {
      return this.replyToRaw ? JSON.parse(this.replyToRaw) : null;
    } catch {
      return null;
    }
  }

  get readByData(): any {
    try {
      return this.readByRaw ? JSON.parse(this.readByRaw) : {};
    } catch {
      return {};
    }
  }

  get deliveredToData(): string[] {
    try {
      return this.deliveredToRaw ? JSON.parse(this.deliveredToRaw) : [];
    } catch {
      return [];
    }
  }

  get deletedForData(): string[] {
    try {
      return this.deletedForRaw ? JSON.parse(this.deletedForRaw) : [];
    } catch {
      return [];
    }
  }

  get mentionsData(): string[] {
    try {
      return this.mentionsRaw ? JSON.parse(this.mentionsRaw) : [];
    } catch {
      return [];
    }
  }
}

File: apps/api/lib/nxtvibes/chats/message.ex
defmodule NxtVibes.Chats.Message do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id
  schema "messages" do
    field :chat_id, :binary_id
    field :chat_type, :string
    field :sender_id, :string
    field :sender_name, :string
    field :text, :string
    field :type, :string
    field :media_url, :string
    field :location, :map
    field :reply_to, :binary_id
    field :status, :string
    field :read_by, :map, default: %{}
    field :delivered_to, {:array, :string}, default: []
    field :deleted_for, {:array, :string}, default: []
    field :deleted_for_everyone_at, :utc_datetime
    field :mentions, {:array, :string}, default: []
    field :edited_at, :utc_datetime

    timestamps(type: :utc_datetime, inserted_at: :created_at, updated_at: :updated_at)
  end

  @doc false
  def changeset(message, attrs) do
    message
    |> cast(attrs, [
      :id,
      :chat_id,
      :chat_type,
      :sender_id,
      :sender_name,
      :text,
      :type,
      :media_url,
      :location,
      :reply_to,
      :status,
      :read_by,
      :delivered_to,
      :deleted_for,
      :deleted_for_everyone_at,
      :mentions,
      :edited_at
    ])
    |> validate_required([:chat_id, :chat_type, :sender_id, :sender_name, :type, :status])
  end
end

